File: src/personalclaw/dashboard/owner_token_url.js
/* The owner-token link, handled in the browser.
 *
 * ONE source, inlined by the gateway (owner_token_url.py) into the three documents that stand
 * between a person and the dashboard: the SPA document a `?token=` link opens, the paste-token
 * Connect gate, and the password sign-in page.
 *
 *   scrub()          takes `token` out of the address bar and the CURRENT history entry, keeping
 *                    the path, every other query parameter and the hash route. By the time a
 *                    document runs, the gateway has already exchanged the token for the HttpOnly
 *                    session cookie, so the URL copy is only something to leak: into a copied
 *                    link, a screenshot, the Back button.
 *   route(hash)      the hash when it is a dashboard route ("#/..."), else "".
 *   connectUrl(raw)  where the Connect gate sends a pasted token: THIS origin's root, the token,
 *                    and the route the gate was opened at. A pasted URL contributes only its
 *                    token, so pasting cannot move anyone to another origin or path.
 *   home()           "/" plus the route — where a successful sign-in lands, instead of "/".
 */
(function (root) {
  'use strict'

  var TOKEN = 'token'
  // "#/" and then only characters a dashboard route uses. The result is only ever the FRAGMENT
  // of a URL built on this origin, so it cannot change where the browser goes; the check keeps
  // it to the shape the hash router reads. Refused: anything that is not "#/...", "#//..."
  // (it reads as a scheme-relative URL), whitespace, control characters, backslashes, and
  // anything longer than a route plausibly is.
  var ROUTE = /^#\/(?!\/)[A-Za-z0-9\-._~!$&'()*+,;=:@%/?]*$/

  function route(hash) {
    var h = String(hash || '')
    return h.length <= 2048 && ROUTE.test(h) ? h : ''
  }

  function scrub() {
    var url = new URL(root.location.href)
    if (!url.searchParams.has(TOKEN)) return false
    url.searchParams.delete(TOKEN)
    var clean = url.pathname + url.search + url.hash
    try {
      root.history.replaceState(root.history.state, '', clean)
    } catch (_) {
      // A webview that refuses History API writes: a replacing navigation has the same
      // property (no history entry keeps the token), and the cookie is already set.
      root.location.replace(clean)
    }
    return true
  }

  function tokenFrom(raw) {
    var value = String(raw || '').trim()
    if (!value) return ''
    try {
      return new URL(value).searchParams.get(TOKEN) || ''
    } catch (_) {
      return value
    }
  }

  function connectUrl(raw) {
    var token = tokenFrom(raw)
    if (!token) return ''
    var target = new URL('/', root.location.origin)
    target.searchParams.set(TOKEN, token)
    target.hash = route(root.location.hash)
    return target.href
  }

  function home() {
    return '/' + route(root.location.hash)
  }

  root.PersonalClawOwnerToken = Object.freeze({
    route: route,
    scrub: scrub,
    connectUrl: connectUrl,
    home: home,
  })

  var self = root.document && root.document.currentScript
  if (self && self.hasAttribute('data-owner-token-scrub')) scrub()
})(window)

File: src/personalclaw/dashboard/owner_token_url.py
"""The owner-token link's browser half — the one script the gateway inlines into its doors.

A fresh browser has no session cookie, so the owner token has to arrive in the first request
(``/?token=…``). The token middleware exchanges it for the HttpOnly ``pc_token_<port>`` cookie on
that response — and the token then stayed in the address bar and the history entry, where a
copied link, a screenshot or the Back button carries it on. And the paste-token Connect gate
rebuilt its target as ``origin + '?token='``, dropping the route the user had opened, so every
deep link landed on the dashboard instead.

``owner_token_url.js`` (package data beside this module) handles both, once: it scrubs the token
with ``history.replaceState``, and it builds the gate's same-origin target keeping only a
validated hash route. It is read ONCE at import, like ``baseline_denylist.json``: a wheel that
lost the file fails to start, instead of serving a Connect gate whose button does nothing.
"""

from __future__ import annotations

from pathlib import Path

#: Present on the inlined tag when the script should scrub the current URL as it runs.
SCRUB_ATTR = "data-owner-token-scrub"

_SOURCE = Path(__file__).with_name("owner_token_url.js").read_text(encoding="utf-8")
if "</script" in _SOURCE.lower():  # pragma: no cover — a packaging guard, not a runtime branch
    raise RuntimeError("owner_token_url.js must not contain a closing script tag")


#: The opening tag of the scrubbing script — what marks a document as already carrying it. The
#: bare attribute name is not enough: the script's own source spells it.
_SCRUB_OPEN = f"<script {SCRUB_ATTR}>"


def script_tag(*, scrub: bool) -> str:
    """The inline ``<script>`` carrying the owner-token helpers (and, with *scrub*, running it)."""
    return f"{_SCRUB_OPEN if scrub else '<script>'}{_SOURCE}</script>"


def script_source() -> str:
    """The bare script body, for a page that composes its own ``<script>`` element."""
    return _SOURCE


def inject_scrub(html: str) -> str:
    """Make the scrubbing script the FIRST thing in ``html``'s ``<head>``.

    First, so the token leaves the address bar before the parser reaches a single resource
    declaration. Idempotent, and a no-op for a document without ``<head>`` — served unchanged
    rather than mangled, like ``surface_layers.inject_safe_meta``.
    """
    if _SCRUB_OPEN in html:
        return html
    head = html.find("<head>")
    if head < 0:
        return html
    at = head + len("<head>")
    return html[:at] + script_tag(scrub=True) + html[at:]
